File: 프로그래머스/3/43162. 네트워크/네트워크.py
from collections import deque

def solution(n, computers):
    parent = [i for i in range(n)]
    
    def find(a):
        if parent[a] == a:
            return a
        parent[a] = find(parent[a])
        return parent[a]
    
    def union(a, b):
        x, y = find(a), find(b)
        
        if x > y:
            parent[x] = y
        elif x < y:
            parent[y] = x
    
    q = deque()
    visited = [False] * n
    
    for i in range(n):
        if visited[i]:
            continue
        q.append(i)
        while q:
            now = q.popleft()
            visited[now] = True
            
            for idx in range(n):
                if idx == now or computers[now][idx] == 0 or visited[idx] == True:
                    continue
                q.append(idx)
                union(i, idx)
        
    #print(parent)
    parent = set(parent)
        
    return len(parent)

File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    
    // bfs
    deque<int> q;
    vector<bool> visited(n, false); // {i번 노드의 방문여부}
    for (int i=0;i<n;i++){
        q.push_back(i);
        if (visited[i]) continue;
        
        while (!q.empty()){
            int now = q.front();
            visited[now] = true;
            q.pop_front();
            for (int next=0;next<n;next++){
                if (next == now || computers[now][next] == 0 || visited[next])
                    continue;
                q.push_back(next);
            }
        }
        answer++;
    }
    
    return answer;
}